Kings County Housing Prices

# Pre-EDA and Data Cleaning

## Importing important libraries

In [1]:
# Import important libraries
import pyforest
import folium
%matplotlib inline

# Hide warnings
import warnings
warnings.filterwarnings("ignore")

# Settings
pd.set_option('display.max_columns', None)

<IPython.core.display.Javascript object>

## Read the Data

The first step is read the data from a csv file to a Pandas DataFrame.

In [2]:
test = pd.read_csv('../data/kc_house_data_test_features.csv', index_col=0)
df = pd.read_csv('../data/kc_house_data_train.csv', index_col=0)

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

## EDA Round 1

In [5]:
# Check the shape of the dataframe
print(df.shape)
# Check the first 5 rows of the train dataframe
df.head()

(17290, 21)


id             date      price  bedrooms  bathrooms  sqft_living  \
0  2591820310  20141006T000000   365000.0         4       2.25         2070   
1  7974200820  20140821T000000   865000.0         5       3.00         2900   
2  7701450110  20140815T000000  1038000.0         4       2.50         3770   
3  9522300010  20150331T000000  1490000.0         3       3.50         4560   
4  9510861140  20140714T000000   711000.0         3       2.50         2550   

   sqft_lot  floors  waterfront  view  condition  grade  sqft_above  \
0      8893     2.0           0     0          4      8        2070   
1      6730     1.0           0     0          5      8        1830   
2     10893     2.0           0     2          3     11        3770   
3     14608     2.0           0     2          3     12        4560   
4      5376     2.0           0     0          3      9        2550   

   sqft_basement  yr_built  yr_renovated  zipcode      lat     long  \
0              0      1986             0    98058  47.4388 -122.162   
1           1070      1977             0    98115  47.6784 -122.285   
2              0      1997             0    98006  47.5646 -122.129   
3              0      1990             0    98034  47.6995 -122.228   
4              0      2004             0    98052  47.6647 -122.083   

   sqft_living15  sqft_lot15  
0           2390        7700  
1           2370        6283  
2           3710        9685  
3           4050       14226  
4           2250        4050

In [4]:
df.dtypes

id                 int64
date              object
price            float64
bedrooms           int64
bathrooms        float64
sqft_living        int64
sqft_lot           int64
floors           float64
waterfront         int64
view               int64
condition          int64
grade              int64
sqft_above         int64
sqft_basement      int64
yr_built           int64
yr_renovated       int64
zipcode            int64
lat              float64
long             float64
sqft_living15      int64
sqft_lot15         int64
dtype: object

In [7]:
# Get information about the data set
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 17290 entries, 0 to 17289
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             17290 non-null  int64  
 1   date           17290 non-null  object 
 2   price          17290 non-null  float64
 3   bedrooms       17290 non-null  int64  
 4   bathrooms      17290 non-null  float64
 5   sqft_living    17290 non-null  int64  
 6   sqft_lot       17290 non-null  int64  
 7   floors         17290 non-null  float64
 8   waterfront     17290 non-null  int64  
 9   view           17290 non-null  int64  
 10  condition      17290 non-null  int64  
 11  grade          17290 non-null  int64  
 12  sqft_above     17290 non-null  int64  
 13  sqft_basement  17290 non-null  int64  
 14  yr_built       17290 non-null  int64  
 15  yr_renovated   17290 non-null  int64  
 16  zipcode        17290 non-null  int64  
 17  lat            17290 non-null  float64
 18  long  

In [9]:
df.describe()

id         price      bedrooms     bathrooms   sqft_living  \
count  1.729000e+04  1.729000e+04  17290.000000  17290.000000  17290.000000   
mean   4.565502e+09  5.407395e+05      3.372470      2.111943   2081.464604   
std    2.874656e+09  3.733190e+05      0.939346      0.770476    920.018539   
min    1.000102e+06  7.500000e+04      0.000000      0.000000    290.000000   
25%    2.114701e+09  3.210000e+05      3.000000      1.500000   1430.000000   
50%    3.903650e+09  4.500000e+05      3.000000      2.250000   1920.000000   
75%    7.301150e+09  6.450000e+05      4.000000      2.500000   2550.000000   
max    9.900000e+09  7.700000e+06     33.000000      8.000000  13540.000000   

           sqft_lot        floors    waterfront          view     condition  \
count  1.729000e+04  17290.000000  17290.000000  17290.000000  17290.000000   
mean   1.524340e+04      1.490312      0.007981      0.238519      3.408502   
std    4.230462e+04      0.538909      0.088985      0.775229      0.651296   
min    5.720000e+02      1.000000      0.000000      0.000000      1.000000   
25%    5.081250e+03      1.000000      0.000000      0.000000      3.000000   
50%    7.642000e+03      1.500000      0.000000      0.000000      3.000000   
75%    1.072575e+04      2.000000      0.000000      0.000000      4.000000   
max    1.651359e+06      3.500000      1.000000      4.000000      5.000000   

              grade    sqft_above  sqft_basement      yr_built  yr_renovated  \
count  17290.000000  17290.000000   17290.000000  17290.000000  17290.000000   
mean       7.654425   1789.306015     292.158589   1970.792019     83.806304   
std        1.174718    829.265107     443.151874     29.343516    400.329376   
min        1.000000    290.000000       0.000000   1900.000000      0.000000   
25%        7.000000   1200.000000       0.000000   1951.000000      0.000000   
50%        7.000000   1560.000000       0.000000   1974.000000      0.000000   
75%        8.000000   2214.500000     560.000000   1996.000000      0.000000   
max       13.000000   9410.000000    4820.000000   2015.000000   2015.000000   

            zipcode           lat          long  sqft_living15     sqft_lot15  
count  17290.000000  17290.000000  17290.000000   17290.000000   17290.000000  
mean   98078.193175     47.560058   -122.214258    1987.986698   12873.475824  
std       53.607949      0.138412      0.140857     684.802635   27227.437583  
min    98001.000000     47.155900   -122.519000     399.000000     659.000000  
25%    98033.000000     47.471200   -122.329000    1490.000000    5111.250000  
50%    98065.000000     47.571600   -122.230000    1840.000000    7622.500000  
75%    98118.000000     47.677900   -122.126000    2360.000000   10101.750000  
max    98199.000000     47.777600   -121.315000    6210.000000  858132.000000

In [11]:
df.isna().any()

id               False
date             False
price            False
bedrooms         False
bathrooms        False
sqft_living      False
sqft_lot         False
floors           False
waterfront       False
view             False
condition        False
grade            False
sqft_above       False
sqft_basement    False
yr_built         False
yr_renovated     False
zipcode          False
lat              False
long             False
sqft_living15    False
sqft_lot15       False
dtype: bool

### Findings:
- The train data set contains 17290 rowns and 22 columns
- There aren't any null values
- Fix the date columns to time series
- Is the date column showing when the house was sold?
- Create columns with years since last renovation
- There is a house with 33 bedrooms. Could be an outlier

## Data Cleaning

### Converting column date to datetime type

In [12]:
# Fix the date column
df['date'] = pd.to_datetime(df['date'])
df.head(2)

<IPython.core.display.Javascript object>

id       date     price  bedrooms  bathrooms  sqft_living  \
0  2591820310 2014-10-06  365000.0         4       2.25         2070   
1  7974200820 2014-08-21  865000.0         5       3.00         2900   

   sqft_lot  floors  waterfront  view  condition  grade  sqft_above  \
0      8893     2.0           0     0          4      8        2070   
1      6730     1.0           0     0          5      8        1830   

   sqft_basement  yr_built  yr_renovated  zipcode      lat     long  \
0              0      1986             0    98058  47.4388 -122.162   
1           1070      1977             0    98115  47.6784 -122.285   

   sqft_living15  sqft_lot15  
0           2390        7700  
1           2370        6283

# EDA

In [26]:
sns.set_style("seaborn")

df.hist(bins=50, figsize=(20,15))
plt.show()

<IPython.core.display.Javascript object>

ValueError: style must be one of white, dark, whitegrid, darkgrid, ticks

### Fix any extremes problems with the data

In [17]:
# Using nlargest to see the rows with the highest number of bedrooms
df.nlargest(5,'bedrooms')

id       date      price  bedrooms  bathrooms  sqft_living  \
8597   2402100895 2014-06-25   640000.0        33       1.75         1620   
2356   1773100755 2014-08-21   520000.0        11       3.00         3000   
3680   8812401450 2014-12-29   660000.0        10       3.00         2920   
11181   627300145 2014-08-14  1148000.0        10       5.25         4590   
17114  5566100170 2014-10-29   650000.0        10       2.00         3610   

       sqft_lot  floors  waterfront  view  condition  grade  sqft_above  \
8597       6000     1.0           0     0          5      7        1040   
2356       4960     2.0           0     0          3      7        2400   
3680       3745     2.0           0     0          4      7        1860   
11181     10920     1.0           0     2          3      9        2500   
17114     11914     2.0           0     0          4      7        3010   

       sqft_basement  yr_built  yr_renovated  zipcode      lat     long  \
8597             580      1947             0    98103  47.6878 -122.331   
2356             600      1918          1999    98106  47.5560 -122.363   
3680            1060      1913             0    98105  47.6635 -122.320   
11181           2090      2008             0    98004  47.5861 -122.113   
17114            600      1958             0    98006  47.5705 -122.175   

       sqft_living15  sqft_lot15  year year_sold mont_sold  
8597            1330        4700  2014      2014   2014-06  
2356            1420        4960  2014      2014   2014-08  
3680            1810        3745  2014      2014   2014-12  
11181           2730       10400  2014      2014   2014-08  
17114           2040       11914  2014      2014   2014-10

Ok, it is an outlier. Based on the price, it sounds a weird that this house contains 33 bedrooms. Luckily, there is an ID that we can use on internet and see if we can find this house.
I was able to find the house with 33 bedrooms on Zillow and Clustmaps. All the information about the house matches, except the number of bedrooms, which seems a typo. The house has 3 bedrooms. Thus, I'll fix it.

In [18]:
# This fixes the house with 33 bedrooms turning it into a 3-bedroom house
df[df['bedrooms'] == 33] = df[df['bedrooms'] == 3]

# This will add 1 bathroom to houses without any bathroom
df['bathrooms'] = df.bedrooms.apply(lambda x: 1 if x < 1 else x)

- The outlier in the bedroom collumn was a typo and it was fixed.
- One bathroom was included to all the houses without bathroom in the data set.
- One bedroom was included to all the houses without a bedroom in the data set.

## Feature Engineering

In [20]:
# Create year and month features
df['year_sold'] = pd.to_datetime(df['date']).dt.to_period('Y')
df['mont_sold'] = pd.to_datetime(df['date']).dt.to_period('M')
df.head()

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

id       date      price  bedrooms  bathrooms  sqft_living  \
0  2.591820e+09 2014-10-06   365000.0       4.0        4.0       2070.0   
1  7.974201e+09 2014-08-21   865000.0       5.0        5.0       2900.0   
2  7.701450e+09 2014-08-15  1038000.0       4.0        4.0       3770.0   
3  9.522300e+09 2015-03-31  1490000.0       3.0        3.0       4560.0   
4  9.510861e+09 2014-07-14   711000.0       3.0        3.0       2550.0   

   sqft_lot  floors  waterfront  view  condition  grade  sqft_above  \
0    8893.0     2.0         0.0   0.0        4.0    8.0      2070.0   
1    6730.0     1.0         0.0   0.0        5.0    8.0      1830.0   
2   10893.0     2.0         0.0   2.0        3.0   11.0      3770.0   
3   14608.0     2.0         0.0   2.0        3.0   12.0      4560.0   
4    5376.0     2.0         0.0   0.0        3.0    9.0      2550.0   

   sqft_basement  yr_built  yr_renovated  zipcode      lat     long  \
0            0.0    1986.0           0.0  98058.0  47.4388 -122.162   
1         1070.0    1977.0           0.0  98115.0  47.6784 -122.285   
2            0.0    1997.0           0.0  98006.0  47.5646 -122.129   
3            0.0    1990.0           0.0  98034.0  47.6995 -122.228   
4            0.0    2004.0           0.0  98052.0  47.6647 -122.083   

   sqft_living15  sqft_lot15  year year_sold mont_sold  
0         2390.0      7700.0  2014      2014   2014-10  
1         2370.0      6283.0  2014      2014   2014-08  
2         3710.0      9685.0  2014      2014   2014-08  
3         4050.0     14226.0  2015      2015   2015-03  
4         2250.0      4050.0  2014      2014   2014-07